ArduCopter/mode_autorotate.cpp

```diff
@@ -0,0 +1,417 @@
+    return false;
+}
+
+    //initialise head speed/collective controller
+    //This must be done before RPM value is fetched
+    arot->init_hs_controller();
+
+    // Retrive rpm and start rpm sensor health checks
+    _initial_rpm = arot->get_rpm(true);
+
+    //Display message 
+    gcs().send_text(MAV_SEVERITY_INFO, "Autorotation initiated");
+
+    //Retrieve parameter values from autorotation library
+    arot->set_param_values(&_param_head_speed_set_point, &_param_accel_max, &_param_target_fwd_speed, &_param_td_alt, &_param_col_entry_cutoff_freq, &_param_col_glide_cutoff_freq, &_param_bail_time);
+
+     //set all inial flags to on
+    _flags.entry_initial = 1;
+    _flags.ss_glide_initial = 1;
+    _flags.flare_initial = 1;
+    _flags.touch_down_initial = 1;
+    _flags.level_initial = 1;
+    _flags.break_initial = 1;
+    _flags.straight_ahead_initial = 1;
+    _flags.bail_out_initial = 1;
+    _msg_flags.bad_rpm = true;
+
+    //Prevent divide by zero error
+    if (_param_head_speed_set_point < 500) {
+        _param_head_speed_set_point = 500;  //Making sure that hover rpm is not unreasonably low
+    }
+
+    //initialise speed/height controller
+    helispdhgtctrl->init_controller();
+
+    //setting default starting switches
+    phase_switch = ENTRY;
+
+    //set entry timer
+    _entry_time_remain = AUTOROTATE_ENTRY_TIME;
+
+    //The decay rate to reduce the head speed from the current to the target
+    _hs_decay = ((_initial_rpm/_param_head_speed_set_point) - HEAD_SPEED_TARGET_RATIO) / AUTOROTATE_ENTRY_TIME;//AUTOROTATE_HS_ADJUST_TIME;
+
+    return true;
+}
+
+
+
+void ModeAutorotate::run()
+{
+
+    //Check if interlock becomes engaged
+    if (motors->get_interlock() && !copter.ap.land_complete) {
+        phase_switch = BAIL_OUT;
+    } else if (motors->get_interlock() && copter.ap.land_complete) {
+        //Aircraft is landed and no need to bail out
+        set_mode(copter.prev_control_mode, MODE_REASON_AUTOROTATION_BAILOUT);
+    }
+
+    //Current time
+    now = millis(); //milliseconds
+
+    //Initialise internal variables
+    float curr_alt = inertial_nav.get_position().z;     // Current altitude
+    float curr_vel_z = inertial_nav.get_velocity().z;   // Current vertical descent
+
+    //----------------------------------------------------------------
+    //                  State machine logic
+    //----------------------------------------------------------------
+
+     //Setting default phase switch positions
+     nav_pos_switch = USER_CONTROL_STABILISED;
+
+    //altitude check to jump to correct flight phase if at low height
+    if (curr_alt <= _param_td_alt) {
+        //jump to touch down phase
+        //phase_switch = TOUCH_DOWN;  //currently switched off in this version.
+    }
+
+    //Timer from entry phase to progress to glide phase
+    if (phase_switch == ENTRY){
+
+        //update entry phase timer
+        _entry_time_remain -= G_Dt;
+
+        if (_entry_time_remain < 0.0f) {
```

Comment:
A more AP standard way of doing this would be to have an _entry_time_start_ms method that is set to AP_HAL::millis() and then have another check which is (AP_HAL::millis - _entry_time_start_ms > XXX).

Reply:
Done as suggested.